File: Leet Code Solutions/1286-constrained-subsequence-sum/1286-constrained-subsequence-sum.js
var constrainedSubsetSum = function(nums, k) {
        let dq = [];
        for (let i = 0; i < nums.length; i++) {
            nums[i] += dq.length > 0 ? nums[dq[0]] : 0;

            while (dq.length > 0 && (i - dq[0] >= k || nums[i] >= nums[dq[dq.length - 1]])) {
                if (nums[i] >= nums[dq[dq.length - 1]]) dq.pop();
                else dq.shift();
            }

            if (nums[i] > 0) {
                dq.push(i);
            }
        }
        return Math.max(...nums);
    }

File: Leet Code Solutions/1286-constrained-subsequence-sum/1286-constrained-subsequence-sum.rs
impl Solution {
    pub fn constrained_subset_sum(mut nums: Vec<i32>, k: i32) -> i32 {
        let mut dq = std::collections::VecDeque::new();
        for i in 0..nums.len() {
            let val_to_add = if let Some(&front) = dq.front() {
                nums[front]
            } else {
                0
            };
            nums[i] += val_to_add;

            while let Some(&front) = dq.front() {
                if i as i32 - front as i32 >= k || nums[i] >= nums[*dq.back().unwrap()] {
                    if nums[i] >= nums[*dq.back().unwrap()] {
                        dq.pop_back();
                    } else {
                        dq.pop_front();
                    }
                } else {
                    break;
                }
            }

            if nums[i] > 0 {
                dq.push_back(i);
            }
        }
        *nums.iter().max().unwrap()
    }
}